### PR TITLE
[Adding llm judge function] Feature/llm judge

### PR DESCRIPTION
**Pull request by pengwei715**
Hi @guy1992l and @ZeevRispler ,

This PR is trying to merge the llm as a judge metrics to the function repo. the basic idea is building three types of metrics.

1. Single grading. The judge will give a score based on the metrics definition, examples, rubric. 
2. Pairwise grading. There will be a bench mark model that will be used to answer the same question. The judge will give two scores to these two answers. This will give you an idea of how your model performs compared with the benchmark model
3. Pairwise grading with reference. Similar to the pairwise grading, for this metrics, we know the ground truth of the answer. The Judge will give two scores with the reference.

For the Judge part, in the code, we support the open source models from huggingface as well as OpenAI API.

Please take a look and let me know your thoughts. Thanks.

### CODE REVIEW

File: llm_judge/llm_judge.py [hunk 1]
```diff
@@ -0,0 +1,1141 @@
...
+                res_dic["score_of_assistant_a"],
+                res_dic["explanation_of_assistant_a"],
+                res_dic["score_of_assistant_b"],
+                res_dic["explanation_of_assistant_b"],
+            ]
+
+        return res_df
+
+
+class OPENAIJudgeSingleGrading(LLMJudgeSingleGrading):
```
Comment (ZeevRispler) [inline llm_judge/llm_judge.py:747]: Do you need to do everything again for OpenAi? no way to integrate it better and re-use some code?

File: llm_judge/test_llm_judge.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    OPENAIJudgePairwiseGrading,
+    OPENAIJudgeReferenceGrading,
+)
+from mlrun.utils import logger
+
+JUDGE_MODEL = "TheBloke/Mistral-7B-OpenOrca-GPTQ"
+JUDGE_CONFIG = {
+    "device_map": "auto",
+    "revision": "gptq-8bit-128g-actorder_True",
+    "trust_remote_code": False,
```
Comment (ZeevRispler) [inline llm_judge/test_llm_judge.py:33]:
```suggestion
```

it's default usually

File: llm_judge/test_llm_judge.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+}
+JUDGE_INFER_CONFIG = {
+    "max_length": 1500,
+}
+TOKENIZER_JUDGE_CONFIG = {"use_fast": True}
+BENCHMARK_MODEL = "microsoft/phi-2"
```
Comment (ZeevRispler) [inline llm_judge/test_llm_judge.py:39]: not sure if best for judge, maybe switch with the mistral model

File: llm_judge/test_llm_judge.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+            Explanation: This answer is completely correct, providing a thorough explanation that adds relevant context about Paris's role as the cultural and political center of France, directly addressing the question with depth and precision.
+                     """,
+    }
+    return prompt_config
+
+
+def test_single_grading_score(prompt_fixture):
+    prompt_config = prompt_fixture
+    q1 = "What is the capital of China?"
+    a1 = "The capital of China is Kongfu"
```
Comment (ZeevRispler) [inline llm_judge/test_llm_judge.py:101]: i don't understand why this is the answer, you want to test with a wrong answer?
(maybe i am missing something, but the answer is beijing as far as i know)

**Review by ZeevRispler — COMMENTED** [review]
great work peng!

File: llm_judge/llm_judge.py [hunk 5]
```diff
@@ -0,0 +1,1141 @@
...
+{rubric}
+You must return the following fields in your output:
+- score of assistant a: a numerical score of {name} for the response
+- explanation of assistant a: a explanation about how did you determine the score step-by-step
+- score of assistant b: a numerical score of {name} for the response
+- explanation of assistant b: a explanation about how did you determine the score step-by-step
+[Output]:
+"""
+
+REF_GRADE_PROMPT = """
```
[the remark this answers is not in the document]
Comment (pengwei715) [reply llm_judge/llm_judge.py:94]: Hi @ZeevRispler , the reference grading is using both bench mark model and ground truth to do the judgement. The idea is if we have some sort of ground truth, we ask the bench mark model to generate the answer, then we give two answers and ground truth to the judge to get an evaluation.

File: llm_judge/llm_judge.py (hunk 1 above)
Comment (pengwei715) [reply llm_judge/llm_judge.py:747]: Looking into it .

File: llm_judge/test_llm_judge.py (hunk 4 above)
Comment (pengwei715) [reply llm_judge/test_llm_judge.py:100]: Hi @ZeevRispler , this is testing if the judge can judge the wrong answer.

**Pull request closed without merging**